### ISSUE TITLE
Reading from Pool is not thread-safe when pool_alloctmpspace() is used

### ISSUE THREAD

**Issue by dmach** [openSUSE/libsolv#471]
Several functions use pool_alloctmpspace() to return a string allocated in Pool's temporary area.
Such strings must be processed and discarded immediately.
Everything works fine as long as the code runs in a single thread.

We're currently writing new dnfdaemon - a DBus interface that processes incoming requests in an infinite loop and turns them into asynchronous threads. We are aware of potential concurrency issues, designing the code to avoid them as much as possible.

Our use case is: load repositories into Pool and let users run multiple queries at a time.
We believe that pool_alloctmpspace() might become an issue when used in threads and that impacts all libsolv functions that use it, incl. pool_solvable2str() for example.

Locking doesn't seem to be a good idea, because a user would have to always unlock it explicitly after the value is consumed and it also decreases performance.
I'm wondering if `thread_local` annotation on `pool->tmpspace` couldn't be used to eliminate race conditions among threads.
Maybe also returning a newly allocated string might be an option, but I don't know the performance impact.
Or do you have any better idea?

**Comment by ignatenkobrain** [comment]
I think more correct approach (which will also make it "safe" for multiple threads, assuming that you don't touch string pool, e.g. by creating new packages) is to mimic pool_solvable2str() yourself and use your temporary memory for those.

I think we could probably have 2 variations of these functions, so that caller can choose which one he wants (thread-safe or fast)…

They way current function is designed is to use as less memory as possible and to be as fast as possible.

**Comment by dmach** [comment]
I don't think it's as easy as you suggested.
`pool->tmpspace` is used across the whole code incl. many lookups.

**Comment by j-mracek** [comment]
@mlschroe I know multi threading is not an easy topic but I would like to ask you for your opinion. We are preparing a new DNF daemon and multi threading support in libsolv is a feature that is critical for the design. Do you think whether is it possible or feasible to make libsolv thread safe? Right now we see an issue with `pool->tmpspace` but there can be other workflows in libsolv that can not be compatible with multi thread usage. Anyway we will try our best to resolve the problem but at beginning we need to understand how big problem is in front of us.

**Comment by mlschroe** [comment]
I don't think you can easily make libsolv multi-thread safe. You can do this for the tmpspace, but there's more to it. For example provider lookup caches the result is the whatprovides hash.

If we want to go this route we need to go through all the API functions and find out if they need locking. The original design was "a pool must only be accessed by one thread at a time", so there might be some ugly corner cases...

**Comment by j-mracek** [comment]
Thank you very much for information. OK, we will try to not use libsolv in multi-thread.
